Integrates AccountService with Frontend.

xiahome opened the pull request:
This PR integrates the AccountService into Frontend, so that target
account and container information are injected into the RestRequest,
which allow access by other consuming components such as ACL service.

Test: ./gradlew clean && ./gradlew build

vgkholla commented on ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java:
```diff
@@ -144,6 +144,16 @@ public Account getAccountById(short id) {
   }
 
+  @Override
+  public short getContainerIdForLegacyPutPublicBlob() {
```
why are these functions and not just static variables in `Account`? Another way of phrasing the question - how does this change across implementations?

vgkholla replied:
Also "Legacy" is too vague a term - its definition changes. Let's give this a rethink - these really shouldn't be function IMO

xiahome replied:
Removed these two methods in the interface. Now just use `Container#UNKNOWN_PUBLIC_CONTAINER_ID` and `Container#UNKNOWN_PRIVATE_CONTAINER_ID` for both `InMemoryUnknownAccountService` and `HelixAccountService`. There should be no case that [-1, 0] and [-1, 1] would be generated, as long as `HelixAccountService` is switched in before turning on BlobId and BlobProperties v2 write path.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/config/HelixPropertyStoreConfig.java:
```diff
@@ -18,6 +18,7 @@
  */
 public class HelixPropertyStoreConfig {
   public static final String HELIX_PROPERTY_STORE_PREFIX = "helix.property.store.";
+  public static final String INVALID_ZK_CLIENT_CONNECT_STRING = "invalidZkClientConnectString";
```
why not just empty string?

xiahome replied:
Changed to using empty string.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -141,6 +153,22 @@
+    /**
+     * The key for the target {@link com.github.ambry.account.Container} indicated by the request.
+     */
+    public final static String TARGET_CONTAINER_KEY = "ambry-internal--key-target-container";
```
typo: two "--" b/w "internal" and "key"

xiahome replied:
Fixed.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -520,6 +541,69 @@ public static String getServiceId(RestRequest restRequest) throws RestServiceExc
+   * @return A boolean to indicate the value of the isPrivate flag.
+   * @throws RestServiceException if exception occurs during parsing the arg.
+   */
+  public static boolean getIsPrivateSetting(Map<String, Object> args) throws RestServiceException {
```
minor: can just name function `isPrivate()`?

xiahome replied:
Renamed.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -520,6 +541,69 @@ public static String getServiceId(RestRequest restRequest) throws RestServiceExc
+   * @throws RestServiceException
+   */
+  public static String getTargetAccountNameFromHeader(RestRequest restRequest) throws RestServiceException {
+    Objects.requireNonNull(restRequest, "restRequest cannot be null");
```
don't need this - the next line will throw a `NullPointerException` if `restRequest` is `null`

xiahome replied:
Removed.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -520,6 +541,69 @@ public static String getServiceId(RestRequest restRequest) throws RestServiceExc
+   * @throws RestServiceException
+   */
+  public static String getTargetContainerNameFromHeader(RestRequest restRequest) throws RestServiceException {
+    Objects.requireNonNull(restRequest, "restRequest cannot be null");
```
same here

xiahome replied:
Removed.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -183,6 +211,14 @@
   public static BlobProperties buildBlobProperties(Map<String, Object> args) throws RestServiceException {
+    Account account = (Account) args.get(InternalKeys.TARGET_ACCOUNT_KEY);
+    Container container = (Container) args.get(InternalKeys.TARGET_CONTAINER_KEY);
+    Objects.requireNonNull(account, "Target account object has not been injected into rest request");
```
I don't think these checks are required - the `NullPointerException` will be thrown down the chain anyway?

xiahome replied:
Removed.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -586,4 +670,22 @@ private static ByteRange buildByteRange(String rangeHeaderValue) throws RestServ
+   * @param isOldValueAllowed {@code true} if an old argument value is allowed; {@code false} otherwise.
+   * @return The old value of the argument.
+   */
+  public static Object setArg(RestRequest restRequest, String key, Object value, boolean isOldValueAllowed) {
```
not sure this function is required. The `requireNonNull` statements look superfluous.

xiahome replied:
Removed.

vgkholla replied:
Is this function required? It looks like a logging function rather than any new functionality. Something that is catching programming errors?

xiahome replied:
Just removed this method from `RestUtils`. Also, reject the requests that have the headers the same as the internal header key.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdV2.java:
```diff
@@ -0,0 +1,45 @@
+  protected short getCurrentVersion() {
+    return BLOB_ID_V2;
+  }
+}
```
minor: new line needed

xiahome replied:
added.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -208,6 +224,17 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
       checkAvailable();
       long propsBuildStartTime = System.currentTimeMillis();
+      accountAndContainerSanityCheck(restRequest);
+      // check if the request is old (with no accountName nor containerName) or new (otherwise)
```
will discuss offline with you if this can be refactored. Prefer to have minimal functionality in this part of the function to make it easier to read.

cgtz replied:
This can probably be moved into a helper method

xiahome replied:
Have wrapped all account/container related logic in one method called `injectAccountAndContainerForPutRequest`

vgkholla commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java:
```diff
@@ -239,6 +240,13 @@ public RestMethod getRestMethod() {
   @Override
+  public Object setArg(String key, Object value) {
+    Object oldVal = allArgs.put(key, value);
+    allArgsReadOnly = Collections.unmodifiableMap(allArgs);
```
why is this step needed? changing the underlying map should have changed this too.

xiahome replied:
Verified that this is not necessary.

vgkholla commented on config/frontend.properties:
```diff
@@ -24,3 +24,10 @@ router.delete.success.target=1
+# helix.property.store.root.path=/ambry/Ambry_Dev/helixPropertyStore
+
+# frontend
+# frontend.account.service.factory=com.github.ambry.account.HelixAccountServiceFactory
```
new line needed.

xiahome replied:
added.

nsivabalan commented on ambry-account/src/main/java/com/github/ambry/account/InMemoryUnknownAccountService.java:
```diff
@@ -20,18 +20,20 @@
   public Account getAccountById(short accountId) {
     checkOpen();
-    return Account.UNKNOWN_ACCOUNT;
+    return accountId == Account.UNKNOWN_ACCOUNT_ID ? Account.UNKNOWN_ACCOUNT : null;
```
Sorry to bring this up again. Can we call it LEGACY_ACCOUNT rather than UNKNOWN_ACCOUNT. There is a chance of confusion between UNKNOWN and INVALID. (Same comment for Container too)

vgkholla replied:
What's invalid? Aren't POST requests rejected if no valid account is present? Legacy is too vague - I think unknown is quite literally the case here, is it not?

xiahome replied:
Just think in this way, `UNKNOWN_ACCOUNT` means the account information is **unknown** when the blob was put.  This `UNKNOWN_ACCOUNT` is really not a legacy account, because there is no such thing. It should not be confusing with invalid, which means the account does not exist.

nsivabalan replied:
ok fine. I buy that.

nsivabalan commented on ambry-account/src/main/java/com/github/ambry/account/InMemoryUnknownAccountService.java:
```diff
@@ -20,18 +20,20 @@
+ * queries by account name to this account service will unconditionally return {@link Account#UNKNOWN_ACCOUNT}. This
+ * account service is in memory, and does not talk to any persistent storage service.
  */
 class InMemoryUnknownAccountService implements AccountService {
```
Have suggested to rename "UNKNOWN_ACCOUNT" to "LEGACY_ACCOUNT" below. On similar lines, would prefer "InMemoryLegacyAccountService".

xiahome replied:
Responded below with the same answer.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdV2.java:
```diff
@@ -0,0 +1,45 @@
+ *  A class that makes {@link BlobId#getCurrentVersion()} to return {@link BlobId#BLOB_ID_V2}, which will serialize
+ *  a blobId to {@link BlobIdV2}.
+ */
+public class BlobIdV2 extends BlobId {
```
why is this required now ?

xiahome replied:
This is for testing the case when get/head/delete blobIds in v2 with meaningful aid and cid. The current blobId does not enable the write path, so not easy to make blobId string in v2.

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1011,143 @@ void markStartTime() {
+    } else {
+      targetContainer = targetAccount.getContainerById(accountService.getContainerIdForLegacyPutPublicBlob());
+    }
+    if (targetContainer == null) {
```
can targetContainer be null incase of targetAccount being non null for old put request?

xiahome replied:
This is not allowed. After injection, there guaranteed to be `account` and `container` objected in the `restRequest`, not `null`

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1011,143 @@ void markStartTime() {
+    if (targetAccount == null) {
+      throw new RestServiceException("Invalid account for putting blob", RestServiceErrorCode.InvalidAccount);
+    }
+    ensureAccountNameMatchOrThrow(targetAccount, restRequest);
```
this seems unwanted. We get targetAccount by fetching account name from rest request in line 1045. So, I don't see a point to check that targetAccount matches with what is found in rest request here.

xiahome replied:
It is possible that `InMemoryUnknownAccountService` will return the `UNKNOWN_ACCOUNT` that does not match the account name specified by the `x-ambry-target-account` header.

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1011,143 @@ void markStartTime() {
+    if (targetContainer == null) {
+      throw new RestServiceException("Invalid container for putting blob", RestServiceErrorCode.InvalidContainer);
+    }
+    ensureContainerNameMatchOrThrow(targetContainer, restRequest);
```
same here

xiahome replied:
Same reason, which is because of the behavior of `InMemoryUnknownAccountService`.

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -700,42 +728,38 @@ public void onCompletion(final GetBlobResult routerResult, Exception routerExcep
-                        }
-                      } else if (!blobNotModified) {
-                        response = routerResult.getBlobDataChannel();
+              (securityResult, securityException) -> {
```
is there any change in this code block apart from using java 8 features ?

xiahome replied:
No, just because I touched this file so I make it J8 look-like.

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1011,143 @@ void markStartTime() {
+
+  /**
+   * Sanity check for {@link RestRequest}. This check ensures that the specified service id, account and container name,
+   * if they exist, should now be the same as the reserved values.
```
typo: " should **not** be the same"

xiahome replied:
Fixed

cgtz commented on ambry-account/src/test/java/com/github/ambry/account/MockHelixAccountServiceFactory.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * A mock implementation of {@link AccountServiceFactory}. This is only for testing purpose and is not thread safe.
+ */
+public class MockHelixAccountServiceFactory extends HelixAccountServiceFactory {
```
Why does this need to extend `HelixAccountServiceFactory`?

xiahome replied:
Because it still needs to go through the constructor of the super class.

cgtz commented on ambry-api/src/main/java/com.github.ambry/config/HelixPropertyStoreConfig.java:
```diff
@@ -18,6 +18,7 @@
  */
 public class HelixPropertyStoreConfig {
   public static final String HELIX_PROPERTY_STORE_PREFIX = "helix.property.store.";
+  public static final String INVALID_ZK_CLIENT_CONNECT_STRING = "";
```
does this need to be public?

xiahome replied:
Yeah, this is checked in `RestServer`

cgtz commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -586,4 +664,20 @@ private static ByteRange buildByteRange(String rangeHeaderValue) throws RestServ
+  public static Object setArg(RestRequest restRequest, String key, Object value, boolean isOldValueAllowed) {
+    Object oldVal = restRequest.setArg(key, value);
+    if (!isOldValueAllowed && oldVal != null) {
+      logger.debug("Unexpected key={} in args. Previous value={} has been cleared.", key, oldVal);
```
This method just debug logs if the arg was previously set?

xiahome commented on ambry-account/src/test/java/com/github/ambry/account/MockHelixAccountServiceFactory.java:
```diff
@@ -0,0 +1,96 @@
+   * @param storeConfig The config for constructing a {@link HelixPropertyStore}.
+   * @return A {@link HelixPropertyStore}.
+   */
+  public HelixPropertyStore getHelixStore(HelixPropertyStoreConfig storeConfig) {
```
It will return an existing store when the `shouldUseMockHelixStore` is set to `true`, but not for the `false` case, which will return a real `HelixPropertyStore` for every call. Have added into the javadoc as you suggest.

xiahome commented on ambry-account/src/test/java/com/github/ambry/account/MockHelixAccountServiceFactory.java:
```diff
@@ -0,0 +1,96 @@
+      helixStore = new MockHelixPropertyStore<>();
+      storeKeyToMockStoreMap.put(storeRootPath, helixStore);
+    }
+    return helixStore;
```
Replaced with this neat expression.

xiahome commented on ambry-api/src/main/java/com.github.ambry/config/HelixPropertyStoreConfig.java:
```diff
@@ -35,9 +36,10 @@
 
   /**
-   * The ZooKeeper server address. This config is required.
+   * The ZooKeeper server address.
```
Done.

xiahome commented:
@vgkholla @cgtz @nsivabalan Have addressed your comments. In addition, 
1. added the logic to make FE to behave the same when using `HelixAccountService`;
2. added logs and metrics.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -56,27 +56,31 @@
+   * but are specified public. {@link #UNKNOWN_PUBLIC_CONTAINER} is one of the containers that use it.
    */
-  public static final short UNKNOWN_PUBLIC_CONTAINER_ID = -2;
+  public static final short UNKNOWN_PUBLIC_CONTAINER_ID = 0;
```
this need not be called "unknown_" anymore - this the default public container id - even for valid accounts. Correct?

xiahome replied:
I am inclined to call it `unknown_`, because the `container` information is **unknown** when the blob was put. However, we have the `isPrivate` flag, so instead of assigning `unknown_container`, we can make it a bit specific, that is `unknown_private_` or `unknown_public_` container. What do you think?

xiahome replied:
Have changed to `default` from offline discussion. The containers are for default public and private puts.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -56,27 +56,31 @@
+   * but are specified private. {@link #UNKNOWN_PRIVATE_CONTAINER} is one of the containers that use it.
    */
-  public static final short UNKNOWN_PRIVATE_CONTAINER_ID = -3;
+  public static final short UNKNOWN_PRIVATE_CONTAINER_ID = 1;
```
same here.

xiahome replied:
Renamed to default

vgkholla commented on ambry-api/src/main/java/com.github.ambry/config/FrontendConfig.java:
```diff
@@ -57,15 +64,27 @@
+  /**
+   * Boolean indicator to specify if unrecognized account and container should be prohibited for old put requests.
+   */
+  @Config("frontend.should.prohibit.old.put.with.unrecognized.account.and.container")
```
frontend.should.discard.invalid.account.container.requests?

nsivabalan replied:
frontend.to.accept/disallow.invalid.account.container.requests?

xiahome replied:
Hard to give a nice name, have changed to `frontend.should.allow.unrecognized.account.and.container.for.pre.containerization.request`

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java:
```diff
@@ -137,6 +137,16 @@
   public final Histogram blobStorageServiceShutdownTimeInMs;
 
+  // Account
+  public final Counter oldPutUnrecognizedAccountCount;
```
lets not have metrics with names like "old" "new" - names are too relative. Let's chat about this so that we have metrics that are useful and long term.

xiahome replied:
Have renamed the metrics. Take a look to see if they could be understandable just by reading their names.
```
  public final Counter unRecognizedServiceIdCount;
  public final Counter unRecognizedAccountNameCount;
  public final Counter missingDefaultContainerCount;
  public final Counter unRecognizedContainerNameCount;
  public final Counter getHeadDeleteUnrecognizedAccountCount;
  public final Counter getHeadDeleteUnrecognizedContainerCount;
  public final Meter putWithServiceIdForAccountNameRate;
  public final Meter putWithAccountAndContainerHeaderRate;
```

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactoryTest.java:
```diff
@@ -41,11 +45,15 @@
   public void getAmbryBlobStorageServiceTest() throws Exception {
     // dud properties. server should pick up defaults
     Properties properties = new Properties();
+    properties.setProperty(HelixPropertyStoreConfig.HELIX_PROPERTY_STORE_PREFIX + "zk.client.connect.string",
```
why are these required ? are they used?

xiahome replied:
This is required when instantiating a `HelixaccountService`. There are two tests using `HelixAccountService`

vgkholla replied:
As discussed offline, this looks unneeded. Let's remove

vgkholla commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java:
```diff
@@ -67,12 +68,12 @@
 
   protected volatile ReadIntoCallbackWrapper callbackWrapper = null;
-  protected volatile Map<String, Object> allArgsReadOnly = null;
+  protected volatile Map<String, Object> allArgsReadOnly = new HashMap<>();
```
why is this instantiated? It will be replaced, right?

vgkholla replied:
As discussed, lets make this `final` and make it wrap `allArgs` right here.

xiahome replied:
sorry, not sure why I got this change, fixed now.

vgkholla commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java:
```diff
@@ -239,6 +240,12 @@ public RestMethod getRestMethod() {
 
   @Override
+  public Object setArg(String key, Object value) {
+    Object oldVal = allArgs.put(key, value);
```
`return allArgs.put(key,value)`

xiahome replied:
Changed as suggested.

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+      }
+    }
+    Container targetContainer;
+    boolean isBlobPrivate = isPrivate(restRequest.getArgs());
```
I see that isPrivate() is parsed to build BlobProperties once again. If its happening in more places, can we deduce it to just to one place and pass the value around

xiahome replied:
It seems a bit unstraigforward and I could not find an easy way to reuse the value, unless we change the `buildBlobProperties` in `RestUtils`.

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+              RestServiceErrorCode.InvalidAccount);
+        } else {
+          logger.warn(
+              "Account cannot be identified for blobId={} with accountId={}. Setting targetAccount to UNKNOWN_ACCOUNT",
```
why is this case allowed ?

nsivabalan replied:
synced up offline. Ignore my comment

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+        }
+      }
+      Container targetContainer = targetAccount.getContainerById(blobId.getContainerId());
+      if (targetContainer == null) {
```
Will come talk to you regarding how GET requests for old blobs will be handled here.

nsivabalan replied:
synced up offline. Ignore my comment

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+    String accountName = getTargetAccountNameFromHeader(restRequest);
+    Account targetAccount = accountService.getAccountByName(accountName);
+    if (targetAccount == null) {
+      frontendMetrics.newPutUnRecognizedAccountCount.inc();
```
lowercase r

xiahome replied:
changed.

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+      throw new RestServiceException("Account cannot be found for accountName=" + accountName + " in new put request",
+          RestServiceErrorCode.InvalidAccount);
+    }
+    ensureAccountNameMatchOrThrow(targetAccount, restRequest);
```
how will the account name not match here?

xiahome replied:
This is only for the case when an `InMemoryUnknownAccountService` is used.

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+          "Container cannot be identified for accountName=" + accountName + " and containerName=" + containerName
+              + " in new put request.", RestServiceErrorCode.InvalidContainer);
+    }
+    ensureContainerNameMatchOrThrow(targetContainer, restRequest);
```
same question here for containers

xiahome replied:
This is not needed, removed.

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+   * @param restRequest The {@link RestRequest} to ensure.
+   * @throws RestServiceException if the {@link Account}'s name does not match the name specified in the {@link RestRequest}.
+   */
+  private void ensureAccountNameMatchOrThrow(Account account, RestRequest restRequest) throws RestServiceException {
```
I think calling this just "ensureAccountNameMatch" is descriptive enough. Same for the method below.

xiahome replied:
Renamed.

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java:
```diff
@@ -137,6 +137,16 @@
 
+  // Account
+  public final Counter oldPutUnrecognizedAccountCount;
+  public final Counter newPutUnRecognizedAccountCount;
```
un-capitalize the R in these metrics

xiahome replied:
Done.

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java:
```diff
@@ -67,13 +67,13 @@
   protected final NettyMetrics nettyMetrics;
-  protected final Map<String, Object> allArgs = new TreeMap<String, Object>(String.CASE_INSENSITIVE_ORDER);
-  protected final Queue<HttpContent> requestContents = new LinkedBlockingQueue<HttpContent>();
+  protected final Map<String, Object> allArgs = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
```
why were the args sorted?

xiahome replied:
Just chat with @vgkholla , this is to use the feature of `TreeMap` to make keys case insensitive. The order does not matter.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -67,6 +76,10 @@
   private static final String OPERATION_TYPE_HEAD = "HEAD";
   private static final String OPERATION_TYPE_DELETE = "DELETE";
   private static final String OPERATION_TYPE_POST = "POST";
+  private static final Set<String> requiredAmbryHeadersForOldPuRequest =
```
Renamed to `requiredAmbryHeadersForPutWithServiceId`. The other one is renamed to `requiredAmbryHeadersForPutWithAccountAndContainerName`

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -67,6 +76,10 @@
   private static final String OPERATION_TYPE_DELETE = "DELETE";
   private static final String OPERATION_TYPE_POST = "POST";
+  private static final Set<String> requiredAmbryHeadersForOldPuRequest =
+      Sets.newHashSet(Headers.AMBRY_CONTENT_TYPE, Headers.SERVICE_ID);
```
Changed as suggested.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -67,6 +76,10 @@
+  private static final Set<String> requiredAmbryHeadersForOldPuRequest =
+      Sets.newHashSet(Headers.AMBRY_CONTENT_TYPE, Headers.SERVICE_ID);
+  private static final Set<String> requiredAmbryHeadersForNewPutRequest =
+      Sets.newHashSet(Headers.AMBRY_CONTENT_TYPE, Headers.TARGET_ACCOUNT_NAME, Headers.TARGET_CONTAINER_NAME);
```
We did, when building blobproperties.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+
+  /**
+   * Injects target {@link Account} and {@link Container} for PUT requests. This method also ensures required headers
+   * are present for old and new PUT requests. A new PUT request is the one that carry both the {@code x-ambry-target-account}
```
Changed.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+  /**
+   * Injects target {@link Account} and {@link Container} for PUT requests. This method also ensures required headers
+   * are present for old and new PUT requests. A new PUT request is the one that carry both the {@code x-ambry-target-account}
+   * and {@code x-ambry-target-container} headers, otherwise it is not a new PUT request.
```
Have rephrased the entire sentence.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+
+  /**
+   * Injects {@link Account} and {@link Container} for the old put request, which does not carry account or container
+   * header.
```
Added.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+    }
+    Container targetContainer;
+    boolean isBlobPrivate = isPrivate(restRequest.getArgs());
+    if (isBlobPrivate) {
```
nice!

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+      targetContainer = targetAccount.getContainerById(Container.UNKNOWN_PUBLIC_CONTAINER_ID);
+    }
+    if (targetContainer == null) {
+      frontendMetrics.oldPutUnRecognizedContainerCount.inc();
```
Changed.

xiahome commented:
@vgkholla @cgtz @nsivabalan Have pushed a new commit to address your comments.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,196 @@ void markStartTime() {
+      frontendMetrics.missingDefaultContainerCount.inc();
+      // @todo The check and config can be removed once every existing ambry user has been migrated to a valid account,
+      // @todo and we want to reject the PUT requests with unidentifiable container name.
+      if (frontendConfig.frontendAllowUnrecognizedAccountAndContainerForPreContainerizationRequest) {
```
this needs to be changed once the config changes as we discussed?

xiahome replied:
Sure, removed.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,196 @@ void markStartTime() {
+      Account targetAccount = accountService.getAccountById(blobId.getAccountId());
+      if (targetAccount == null) {
+        frontendMetrics.getHeadDeleteUnrecognizedAccountCount.inc();
+        // @todo The check can be removed once HelixAccountService is running with UNKNOWN_ACCOUNT created.
```
can we assume this will happen and remove this piece of code?

xiahome replied:
We can, but not for every ambry user who change to containerization. For now I prefer to keep this code to allow the situation that Ambry starts before `UNKNOWN_ACCOUNT` is added into AccountService.

xiahome commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -520,6 +538,65 @@ public static String getServiceId(RestRequest restRequest) throws RestServiceExc
+   * accountName was set in the request.
+   * @throws RestServiceException
+   */
+  public static String getTargetAccountNameFromHeader(RestRequest restRequest) throws RestServiceException {
```
Removed and now using the `RestUtils` method. Initially just want this to be the same as the way we get `serviceId`.

xiahome replied:
Also removed the method for getting service id.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,196 @@ void markStartTime() {
+      }
+      setTargetAccountAndContainerInRestRequest(restRequest, targetAccount, targetContainer);
+    } catch (RestServiceException e) {
+      throw e;
```
throwing, catching and rethrowing exceptions is a bit weird. Can we have the`try..catch` around code that doesn't throw `Exception`? (if that doesn't create readability issues).

xiahome replied:
Refactored.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -41,14 +45,20 @@
 import com.github.ambry.router.ReadableStreamChannel;
 import com.github.ambry.router.Router;
 import com.github.ambry.router.RouterException;
+import com.google.common.collect.ImmutableSet;
```
Changed as suggested.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -88,23 +102,23 @@
-   * @param securityServiceFactory the {@link SecurityServiceFactory} to use to get an {@link SecurityService}.
    * @param clusterMap the {@link ClusterMap} in use.
+   * @param idConverterFactory the {@link IdConverterFactory} to use to get an {@link IdConverter} instance.
+   * @param securityServiceFactory the {@link SecurityServiceFactory} to use to get an {@link SecurityService} instance.
```
Added.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -208,6 +226,7 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
       logger.trace("Handling POST request - {}", restRequest.getUri());
       checkAvailable();
       long propsBuildStartTime = System.currentTimeMillis();
+      injectAccountAndContainerForPutRequest(restRequest);
```
Renamed.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -995,4 +1003,196 @@ void markStartTime() {
+    boolean isBlobPrivate = isPrivate(restRequest.getArgs());
+    Container targetContainer = targetAccount.getContainerById(
+        isBlobPrivate ? Container.DEFAULT_PRIVATE_CONTAINER_ID : Container.DEFAULT_PUBLIC_CONTAINER_ID);
+    if (targetContainer == null) {
```
That would be a misconfiguration. Do you think we should still keep this check?

xiahome commented:
@vgkholla @cgtz Have responded to the comments.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java:
```diff
@@ -152,7 +151,7 @@
    */
   public FrontendMetrics(MetricRegistry metricRegistry) {
-    // RestRequestMetrics instances
+    // RestRequestMetrics instancesAmbryBlobStorageServiceTest
```
seems like a stray change

xiahome replied:
Removed.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -1011,16 +1016,23 @@ void markStartTime() {
       ensureRequiredHeadersOrThrow(restRequest, requiredAmbryHeadersForPutWithAccountAndContainerName);
       frontendMetrics.putWithAccountAndContainerHeaderRate.mark();
       injectAccountAndContainerUsingAccountAndContainerHeaders(restRequest);
     } else {
```
this can be `else if` and `else`

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -1104,44 +1091,43 @@ private void injectAccountAndContainerUsingAccountAndContainerHeaders(RestReques
+      blobId = new BlobId(blobIdStr, clusterMap);
     } catch (Exception e) {
       throw new RestServiceException(
           "Invalid blob id=" + blobIdStr + " during injecting account and container from blobId string",
```
minor: think you can drop the ".. during injecting ..." and just say invalid blob id and the id

xiahome replied:
Removed the second half message.

xiahome commented:
@vgkholla Have updated the unit tests.

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -1268,6 +1480,266 @@ private JSONObject createRequestHeaders(ByteRange range, GetOption getOption) th
+      try {
+        doOperation(restRequest, restResponseChannel);
+        if (expectedRestErrorCode != null) {
+          if (expectedRestErrorCode != null) {
```
seems like if check has been made twice by mistake

xiahome replied:
Removed the second check.

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -1658,3 +2135,65 @@ public void close() {
+
+  @Override
+  public AccountService getAccountService() throws InstantiationException {
+    if (accountService == null) {
```
This can be simplified I think
```
if(accountService == null) {
  accountService = new FrontendTestAccountService();
}
return accountService;
```

cgtz replied:
Is the factory required here? We pass the `AccountService` directly to the `BlobStorageService`, right?

xiahome replied:
You are right, this factory can be removed.

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -104,23 +121,74 @@
+  }
+
+  /**
+   * Running for both {@link FrontendConfig#frontendAllowServiceIdBasedPostRequest} is {@code true} and {@code false}.
```
I don't think the parametrization is justified here

xiahome replied:
Have refactored the tests and removed parameterization.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -56,44 +56,50 @@
    * The name of {@link #UNKNOWN_CONTAINER}.
    */
-  public static final String UNKNOWN_CONTAINER_NAME = "ambry-unknown-container-name";
+  public static final String UNKNOWN_CONTAINER_NAME = "ambry-unknown-container";
```
why do we need prefix of "ambry" in the values? Its understood that "Accounts" and "Containers" makes sense only for Ambry.

xiahome replied:
Just introduce a name space to make it less possible to collide with other account and container names.

nsivabalan commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -1019,4 +1031,173 @@ void markStartTime() {
+        targetAccount = Account.UNKNOWN_ACCOUNT;
+      }
+    }
+    Container targetContainer = targetAccount.getContainerById(blobId.getContainerId());
```
this targetContainer is expected to contain UNKNOWN_CONTAINER for old blobIds right ? Where exactly the public or private container is chosen ?

xiahome replied:
This is not the place to choose public or private container. It should be after getting blob info from server, which has not been implemented yet.

nsivabalan commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -1067,7 +1276,8 @@ private void verifyOperationsAfterDelete(String blobId, JSONObject expectedHeade
     GetOption[] options = {GetOption.Include_Deleted_Blobs, GetOption.Include_All};
     for (GetOption option : options) {
-      getBlobAndVerify(blobId, null, option, expectedHeaders, expectedContent);
+      getBlobAndVerify(blobId, null, option, expectedHeaders, expectedContent, Account.UNKNOWN_ACCOUNT,
```
can blobId have valid account and container ?

xiahome replied:
It can, and these cases should be covered in `injectionAccountAndContainerForGetHeadDeleteTest`.

nsivabalan commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -1268,6 +1480,266 @@ private JSONObject createRequestHeaders(ByteRange range, GetOption getOption) th
+          restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_ACCOUNT_KEY));
+      assertEquals("Wrong container object in RestRequest's args", expectedContainer,
+          restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_CONTAINER_KEY));
+      assertEquals("Wrong account object in RestRequest's args", expectedAccount,
```
why repeated verification?

xiahome replied:
My hand shakes when ctrl c&p

nsivabalan commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -1268,6 +1480,266 @@ private JSONObject createRequestHeaders(ByteRange range, GetOption getOption) th
+   * @return The blobId string if the put operation is successful, {@link null} otherwise.
+   * @throws Exception
+   */
+  private String putBlobAndVerifyWithAccountAndContainer(String accountName, String containerName, String serviceId,
```
didn't we have a similar method before w/o account and container ? Why a completely new method ?

xiahome replied:
I think this is a valid point. I was trying to reuse `postBlobAndVerify`, but it seems I will still go through similar amount of code changes because of the error code check logic. So I would prefer to keep this method for now, which stay logically isolated (This method does not check other fields).

vgkholla replied:
can you name it `postBlob()` please?

xiahome replied:
Renamed.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -104,23 +121,74 @@
+    boolean refContainerPrivacy = random.nextBoolean();
+    refContainer = new ContainerBuilder(refContainerId, refContainerName, refContainerStatus, refContainerDescription,
+        refContainerPrivacy, refAccountId).build();
+    refDefaultPublicContainer =
```
We cannot, coz these containers are created per account. Their `parentAccountId`s are different.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -104,23 +121,74 @@
+    short refAccountId = Utils.getRandomShort(random);
+    String refAccountName = UtilsTest.getRandomString(10);
+    Account.AccountStatus refAccountStatus =
+        random.nextBoolean() ? Account.AccountStatus.ACTIVE : Account.AccountStatus.INACTIVE;
```
Set to `ACTIVE`.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -87,15 +96,23 @@
+@RunWith(Parameterized.class)
 public class AmbryBlobStorageServiceTest {
-
+  private static final Random random = TestUtils.RANDOM;
```
Removed the local variable.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -104,23 +121,74 @@
+    short refContainerId = Utils.getRandomShort(random);
+    String refContainerName = UtilsTest.getRandomString(10);
+    Container.ContainerStatus refContainerStatus =
+        random.nextBoolean() ? Container.ContainerStatus.ACTIVE : Container.ContainerStatus.INACTIVE;
```
Changed to `ACTIVE`

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -79,34 +90,42 @@
     this.idConverterFactory = idConverterFactory;
     this.securityServiceFactory = securityServiceFactory;
-    this.clusterMap = clusterMap;
+    this.accountService = accountService;
```
Yes, should  I add `requireNotNull`?

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -399,6 +469,113 @@ public void postGetHeadDeleteTest() throws Exception {
+   * Tests injecting target {@link Account} and {@link Container} for PUT requests. The {@link AccountService} is
+   * prepopulated with a reference account and {@link Account#UNKNOWN_ACCOUNT}. The expected behavior should be:
+   *
+   * accountHeader    containerHeader   serviceIdHeader     expected Error      injected account      injected container
```
done.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -399,6 +469,113 @@ public void postGetHeadDeleteTest() throws Exception {
+   * Tests injecting target {@link Account} and {@link Container} for GET/HEAD/DELETE requests. The {@link AccountService}
+   * is prepopulated with a reference account and {@link Account#UNKNOWN_ACCOUNT}. The expected behavior should be:
+   *
+   * AId in blobId    CId in blobId     expected Error      injected account      injected container
```
Done.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -665,17 +852,25 @@ private RestRequest createRestRequest(RestMethod restMethod, String uri, JSONObj
    * @param contentType sets the {@link RestUtils.Headers#AMBRY_CONTENT_TYPE} header. Required and has to be a valid MIME
    *                    type.
    * @param ownerId sets the {@link RestUtils.Headers#OWNER_ID} header. Optional - if not required, send null.
+   * @param targetAccountName sets the {@link RestUtils.Headers#TARGET_ACCOUNT_NAME} header.
```
Added.

xiahome commented on ambry-account/src/test/java/com/github/ambry/account/MockHelixAccountServiceFactory.java:
```diff
@@ -0,0 +1,93 @@
+   * @param storeConfig The config for constructing a {@link HelixPropertyStore}.
+   * @return A {@link HelixPropertyStore}.
+   */
+  public HelixPropertyStore getHelixStore(HelixPropertyStoreConfig storeConfig) {
```
Coz this is called from `ambry-frontend`

xiahome commented on ambry-api/src/test/java/com.github.ambry/rest/RestUtilsTest.java:
```diff
@@ -674,6 +661,8 @@ private void setAmbryHeadersForPut(JSONObject headers, String ttlInSecs, String
   private void verifyBlobPropertiesConstructionSuccess(JSONObject headers) throws Exception {
     RestRequest restRequest = createRestRequest(RestMethod.POST, "/", headers);
+    restRequest.setArg(RestUtils.InternalKeys.TARGET_ACCOUNT_KEY, Account.UNKNOWN_ACCOUNT);
+    restRequest.setArg(RestUtils.InternalKeys.TARGET_CONTAINER_KEY, Container.UNKNOWN_CONTAINER);
```
Thanks for catching this, added.

xiahome commented on ambry-rest/src/main/java/com.github.ambry.rest/RestServer.java:
```diff
@@ -170,9 +173,14 @@ public RestServer(VerifiableProperties verifiableProperties, ClusterMap clusterM
+    HelixPropertyStoreConfig helixStoreConfig = new HelixPropertyStoreConfig(verifiableProperties);
+    Notifier notifier = null;
+    if (helixStoreConfig.zkClientConnectString != HelixPropertyStoreConfig.INVALID_ZK_CLIENT_CONNECT_STRING) {
+      notifier = new HelixNotifier(helixStoreConfig);
```
This is a tricky part. Besides `HelixAccountService`, we provide `InMemoryUnknownAccountService` for those Ambry adopters who do not want to migrate to containerization. The setup using `InMemoryUnknownAccountService` does not require a `notifier`, so it can be null. And instantiating a `notifier` depends on if we supply the zk config, that is why the check is here.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -399,6 +469,113 @@ public void postGetHeadDeleteTest() throws Exception {
+   * is prepopulated with a reference account and {@link Account#UNKNOWN_ACCOUNT}. The expected behavior should be:
+   *
+   * AId in blobId    CId in blobId     expected Error      injected account      injected container
+   * realAId           realCId          NotFound            refAccount            refContainer
```
Both `account` and `container` are valid, so the `account` and `container` checks are passed. However the blobId is hand-made and does not exist in the router, so it returns `NotFound`.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -399,6 +469,113 @@ public void postGetHeadDeleteTest() throws Exception {
+    verifyAccountAndContainerFromBlobId(blobId, null, null, RestServiceErrorCode.InvalidContainer);
+
+    // aid=refAId, cid=nonExistCId
+    blobId = new BlobIdV2(BlobId.DEFAULT_FLAG, ClusterMapUtils.UNKNOWN_DATACENTER_ID, refAccount.getId(), (short) 1234,
```
Good point, now changed to a negative value.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -399,6 +469,113 @@ public void postGetHeadDeleteTest() throws Exception {
+    verifyAccountAndContainerFromBlobId(blobId, null, null, RestServiceErrorCode.InvalidContainer);
+
+    // aid=nonExistAId, cid=refCId
+    blobId =
```
Same thing done.

xiahome commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -399,6 +469,113 @@ public void postGetHeadDeleteTest() throws Exception {
+    verifyAccountAndContainerFromBlobId(blobId, null, null, RestServiceErrorCode.InvalidAccount);
+
+    // aid=nonExistAId, cid=nonExistCId
+    blobId = new BlobIdV2(BlobId.DEFAULT_FLAG, ClusterMapUtils.UNKNOWN_DATACENTER_ID, (short) 1234, (short) 11,
```
Same thing doen.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -1019,4 +1031,173 @@ void markStartTime() {
+      throw new RestServiceException("Invalid account for putting blob", RestServiceErrorCode.InvalidAccount);
+    }
+    String targetContainerName = getHeader(restRequest.getArgs(), Headers.TARGET_CONTAINER_NAME, false);
+    if (Container.UNKNOWN_CONTAINER_NAME.equals(targetContainerName) || Container.DEFAULT_PUBLIC_CONTAINER_NAME.equals(
```
default public and private are not invalid containers - lets remove this check and add tests that check that blobs can be put into these containers.

xiahome replied:
Removed checking for the two default container names.

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -1019,4 +1031,173 @@ void markStartTime() {
+  private void ensureAccountNameMatch(Account account, RestRequest restRequest) throws RestServiceException {
+    String accountNameFromHeader = getHeader(restRequest.getArgs(), Headers.TARGET_ACCOUNT_NAME, false);
+    if (accountNameFromHeader != null && !accountNameFromHeader.equals(account.getName())) {
+      throw new RestServiceException("Invalid account for putting blob", RestServiceErrorCode.InvalidAccount);
```
lets make this an internal server error

xiahome replied:
Changed.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -79,34 +90,42 @@
   private boolean isUp = false;
 
+  static {
+    requiredAmbryHeadersForPutWithServiceId =
```
Remvoed.

xiahome commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -135,6 +154,9 @@ public void shutdown() {
         idConverter.close();
         idConverter = null;
       }
+      if (accountService != null) {
```
Added a null check.

vgkholla commented:
LGTM
